cmd/kaf/consume.go

```diff
@@ -4,63 +4,113 @@ import (
 	"bytes"
 	"encoding/binary"
 	"fmt"
+	"math"
 	"os"
+	"sort"
 	"strconv"
+	"strings"
 	"sync"
+	"sync/atomic"
 	"text/tabwriter"
 	"time"
 
 	"github.com/Shopify/sarama"
 	"github.com/birdayz/kaf/avro"
 	"github.com/birdayz/kaf/pkg/proto"
 	"github.com/golang/protobuf/jsonpb"
-	prettyjson "github.com/hokaccha/go-prettyjson"
-	colorable "github.com/mattn/go-colorable"
+	"github.com/hokaccha/go-prettyjson"
+	"github.com/mattn/go-colorable"
 	"github.com/spf13/cobra"
 )
 
 var (
-	offsetFlag  string
-	raw         bool
-	follow      bool
-	schemaCache *avro.SchemaCache
-	keyfmt      *prettyjson.Formatter
+	partitionsWhitelistFlag []int
+	offsetFlags             []string
+	countFlags              []int
+	endFlag                 bool
+	raw                     bool
+	followFlag              bool
+	tailFlag                int
+	schemaCache             *avro.SchemaCache
+	keyfmt                  *prettyjson.Formatter
 
 	protoType string
 )
 
 func init() {
 	rootCmd.AddCommand(consumeCmd)
-	consumeCmd.Flags().StringVar(&offsetFlag, "offset", "oldest", "Offset to start consuming. Possible values: oldest, newest.")
-	consumeCmd.Flags().BoolVar(&raw, "raw", false, "Print raw output of messages, without key or prettified JSON")
-	consumeCmd.Flags().BoolVarP(&follow, "follow", "f", false, "Shorthand to start consuming with offset HEAD-1 on each partition. Overrides --offset flag")
-	consumeCmd.Flags().StringSliceVar(&protoFiles, "proto-include", []string{}, "Path to proto files")
-	consumeCmd.Flags().StringSliceVar(&protoExclude, "proto-exclude", []string{}, "Proto exclusions (path prefixes)")
-	consumeCmd.Flags().StringVar(&protoType, "proto-type", "", "Fully qualified name of the proto message type. Example: com.test.SampleMessage")
+	consumeCmd.Flags().StringSliceVarP(&offsetFlags, "offset", "o", []string{},
+		`Offset to start consuming. Pass a single value to apply to all partitions; pass a comma-delimited list to apply to specific partitions.
+Possible values: oldest, newest, -X, +X, X. (X is a number)
```

Comment:
i am a bit skeptical about the -X +X syntax. it does not feel "natural" to me; instead this functionality should be moved into separate flags. i don't want users to learn specific syntaxes for flags - that's one reason why i made kaf, since others tools; e.g. kafkacat and kt where too annoying to learn there imho

Reply:
Any suggestions for what the separate flags should be called? I'm having a hard time thinking of good ones. Here are my thoughts:

Must keep `--offset [oldest|newest|#]` for backwards compatibility.

It sounds like your top priority is making the syntax intuitive, so it's easy to learn and remember. In my experience using my patch for a few weeks it's hard for my fingers to remember that I need to add the `-` if I want to start near the end of the partition, so I often write `kaf consume topic -o 10` and then have to figure out what is going on when I get tons of messages instead of the last 10 from each partition. On the flip side, when I realize my mistake it's easy to remember what I should have written, but I think you're right about +X/-X being unnatural.

I think `--skip` is pretty natural flag for the meaning "start at min offset + X", but it has a slight connotation of "skip the first X messages" which might be confusing.

I'm less sure about the flag meaning "start at max offset - X". `--tail` is almost right, but to me it has very strong connotations of "take the last X items".

How about having flags like this:
`-s, --skip int`:  Skip the first X messages by dividing X by number of partitions
`-S, --skips ints`: Skip the first X messages on each partition, or skip X,Y,Z messages on partitions 0,1,2
`-t, --tail int`: Take the last X messages by dividing X by number of partitions
`-T, --tails ints`:  Take the last X messages on each partition, or the last X,Y,Z messages on partitions 0,1,2 

My most common use case is that I want to see the most recent messages on a topic, without regard to number of partitions, so I made that the more natural case. Do you think that makes sense?

Reply:
I have given this some thought, and this is also the reason i'm slow responding to this. To be honest, i like this even less. I think your initial proposal is the better choice:

1. Allow - and + for --offset, to indicate that you want a relative offset to start (+) or end (-)
2. The notation you suggested for providing multiple partitions is good. In any case, it should be optional, if no specific partition info is given, it's applied for all partition (this is what you did iirc)
3. -e flag is fine. to be honest, i am torn between having -f mandatory to follow, but i'm not willing to break the API. so yeah, the -e is good.

sorry for partly wasting your time, but i think it was necessary to get a feeling for it.

Reply:
I would like to see the original syntax restored. It was more intuitive, but thought it could use some tweaks. My thoughts:
1. Offset syntax should follow the `tail -n` syntax. So the syntax would be `--offset [+]NUM`. Although, there is potential for a race condition when using the positive offset. For example, if we use `--offset +10` and grab the offset, but before we start consuming messages start dropping due to the retention period.
2. Syntax looks good.
3. Making -f mandatory to follow would adhere to standard conventions (e.g. docker logs, tail) and removes the need for endFlag. Could do a major version bump or look to update it on the next major version bump.